### DISCUSSION TITLE
How to build for win arm64?

### DISCUSSION THREAD

**Discussion by timoreichen** [shannah/jdeploy#123] [Q&A] [answered]
Thought i just try:

`"bundles": [
            "mac-x64",
            "mac-arm64",
            "win",
            "win-arm64",
            "linux"
        ],`

But no special download available. arm64, for windows 11 on parallels on a mac m1,2,3,4?

**Comment by shannah** [comment]
Right now I don't have arm builds of windows or linux.  This is on my list but it requires me to either figure out how to cross-compile on an x64 github runner, or set up a self-hosted github runner on arm64.

What kinds of machines run windows on arm64 currently?  Most/all of the consumer desktops and laptops I see are on x64.

**Reply by timoreichen** [reply]
Yes, you are totally right. I was not sure, if i missed that in the docs.. I use it myself mainly now (parallels on a m1 mac). While testing out my jdeploy'ed javafx app i see its a lot slower on virtualized Windows 11. Which gives me a bit of a bad feeling, as i cant really test it out in the same speed as on mac. Well, that would not be a reason to implement it as its "ok" to use it - although slower. But I want to deploy the app to a Microsofts Surface's (a maybe user..) - which is also is x86. Fine too. But i have a feeling that this company will have **Surface Pro X**, which is arm and i think the only one, very soon..

**Comment by timoreichen** [comment]
Found this: https://bugs.openjdk.org/browse/JDK-8314064. I dont now if that helps, but it should work at least without cross-compilation on javafx 22.

**Reply by timoreichen** [reply]
Depending on the fixed issue above it should be either possible. Building which i think is not what you are able to to now but crosscompiling should be possible already.

Taking that:
https://stackoverflow.com/questions/75006480/javafx-maven-platform-specific-build-mac-aarm64-qualifier

cross-compiling should be as easy as do:

mvn package -Djavafx.platform=windows-aarch64

I dont now if this helps for the githubrunner though.

**Comment by reportmill** [comment]
So it looks like this is supported now - my app runs on Surface Pro ARM and in Parallels on Mac M4. However, it looks like it downloads Java 11, even though I need Java 8. My SnapCode needs Java 8 for now (I'm limited in the browser by CheerpJ) and Java 11 is incompatible with my tools.jar, which is stuck on Java 8.

I'm guessing though that there might not be a Java 8 for Windows ARM, so there might not be a solution here (though I'm really hoping for CheerpJ to get me to Java 11 or above :-) )

**Reply by shannah** [reply]
Do you have `javaVersion` set to "8" in the package.json file?

**Reply by reportmill** [reply]
Well, I did find an Oracle JRE 8 for Windows ARM and I installed that and used it to run the jdeploy jar. I tried adding a link to it in ~/.jdeploy/jre-amd64/8/jre/jdk8 (points to \Program Files\Java\jdk-1.8). But it still uses the jre11.

**Reply by reportmill** [reply]
Yep, I have it set to javaVersion 8. It looks like there is no JRE 8 Zulu though (so I assume it just defaulted to jre 11 instead).

**Reply by shannah** [reply]
Which app are you trying this on?  I'd like to try and install to take a look.
I do recall some issues with windows on java 8, but I don't remember what it was, or how I solved it.

**Reply by reportmill** [reply]
This is ~snapcodejava, running on Windows ARM. The app runs fine, but if you click on one of the samples (like Tetris), it will try and fail to compile (because Java 11 doesn't like my tools.jar from Java 8).

**Reply by shannah** [reply]
Off to a dance recital.  Will look later today.

One thing to note is that i have not compiled the launcher for arm yet so if it is running on arm, that is news to me.  probably it's running on some sort of emulation.

**Reply by reportmill** [reply]
That's pretty amazing - because it seems to be running perfectly, even downloaded the jre-amd64.

**Reply by shannah** [reply]
Found the issue with java 8.  I posted a ticket here  https://github.com/shannah/jdeploy/issues/131

I'm close to finding a solution.  Just need to bang my head against it for a while longer.

**Comment by timoreichen** [comment]
Hi. Maybe any news on that...?

**Reply by shannah** [reply]
Thanks for the nudge.  i see that arm runners are now available on github, so i should be able to do this now.

https://github.blog/changelog/2024-09-03-github-actions-arm64-linux-and-windows-runners-are-now-generally-available/

Not sure when i'll have a chance.  Hopefully within next few weeks.

**Reply by timoreichen** [reply]
That would be really helpful and thus fantastic! Thanks.

**Comment by GCalToolkit** [comment]
+1 for the ARM builds! The x86_64 installers are working fine without any issues on ARM Windows 11 in a UTM VM (emulation?), but ARM Linux (Ubuntu 22.04) errors out with "package architecture (amd64) does not match system (arm64)".

JDeploy is such a great tool, thank you :-)

**Comment by reportmill** [comment]
I totally agree - JDeploy is amazing! I don’t know what I would do without it - things had gotten so bad with my old native app stuff, I had mostly given up. Thanks Steve!

jeff

**Comment by shannah** [comment]
Today, I released the new jDeploy desktop app which aims to make it easier for new users to get started.  It is deployed using jDeploy as a desktop app.   No longer requires _any_ dependencies (like npm).  It provides gui options for creating new projects from templates (@reportmill, it would be cool to add a project template for snapcode based projects), and configuring existing projects to us jDeploy.

It supports publishing to both npm and GitHub directly in the app (GitHub used to only work using github actions - now it works in app by pressing "Publish").  It also fixes up the authentication workflow for npm and GitHub with support for 2FA.

I'll be writing a blog post about this to announce it, but thought I'd drop it here as a glimpse of what I've been working on lately.

Now that this is out the door, I think the arm64 builds is my next highest priority.

The [manual ](https://www.jdeploy.com/docs/manual/) has been updated with the new instructions for the GUI app, and the [homepage ](https://www.jdeploy.com/) has a new "[Download](https://github.com/shannah/jdeploy-desktop-gui/releases/tag/master)" link.

**Comment by reportmill** [comment]
Very cool! It never occurred to me to wonder why jDeploy didn’t use jDeploy. :-)

I’ll start looking into adding template support to SnapCode. Being able to quickly share a project is the holy grail, though I have to admit I don’t even have support in SnapCode to build a stand-alone jar (though it’s been on my mind and shouldn’t be hard).

I gave the new app a try, but hit a little npm issue. It asked for my npm credentials and I think I goofed it up (it said it published to npm, but I don’t think it did since I got no email confirmation). Then I tried to turn off npm publishing, but it kept trying the npm step. My command-line workflow still works fine.

jeff

**Reply by shannah** [reply]
Hmm.  Can you post a screen shot of the publish dialog when it says that it published to npm?

**Reply by shannah** [reply]
Wait.  I think I see what happened.  I was using an AI tool to help merge some of my new documentation into the manual, and it looks like it goofed some stuff up.  It completely stripped out the section on setting up the npm token.

I have reverted the change, and manually merged in the new "Getting Started" documentation.

The Gist: Via the GUI you need to use a token for NPM publishing.  You don't use your password.  I really tried to get it working with the password, but they don't offer an API for this, nor do they encourage it, so all of my attempts were flakey.  The token approach is stable and supported.  Just requires that first step of generating a token one time.

You can see the instructions for setting up the token here:  https://www.jdeploy.com/docs/manual/#_generating_an_npm_token

**Comment by reportmill** [comment]
Below is what it says (I tried again this morning, and entered a proper npm token). I also tried to turn off the npm switch, but it seems to have no effect (even if I quit and restart the app after turning the switch off).

Using command [/Users/jeff/.jdeploy/node/v23.1.0/bin/npm, publish]
Package published to npm successfully.
Waiting for npm to update its registry...
Uploading icon to jdeploy.com...
Upload complete
There was a problem publishing the icon to https://www.jdeploy.com/
Error message: Failed load package from npm registry. Version 2025.3.16 wasn't found in package snapcodejava

**Reply by shannah** [reply]
Does it say anything in your `~/.jdeploy/log/ jdeploy-desktop-gui.lastrun.log` file?

**Reply by reportmill** [reply]
There is no `~/.jdeploy/log/jdeploy-desktop-gui.lastrun.log` but there is a `/.jdeploy/log/github.com/shannah/jdeploy-desktop-gui/jdeploy-desktop-gui.lastrun.log`, but it is zero bytes (and from  yesterday, not today).

**Reply by shannah** [reply]
Okay, I'll file this under "I don't understand how it ever worked at all".
Just bought a new Mac, and was able to reproduce your issue first time, @reportmill .  Probably it worked for me in testing because I had node installed on the machine, whereas with the jdeploy desktop, it bundles its own node, so you shouldn't need it installed.

In any case, I have fixed it.  If you open it and try to publish again, it should work, as it will download the latest fixes automatically.

**Reply by shannah** [reply]
One other note.  You'll probably need to first delete the `~/.jdeploy/node` directory to force it to download node again.

**Comment by reportmill** [comment]
Very nice - works perfectly for me now!

Just for fun, I tried to turn off the “npm” switch but it still does the npm publish (which doesn’t effect me because I think I want to keep doing npm for now).

Thanks again!

jeff

**Comment by shannah** [comment, marked as answer]
Note: v5.0 now includes support for Arm64 builds on Windows and linux.

https://www.jdeploy.com/docs/releases/5.0/release-notes.html
